libs/core/langchain_core/runnables/context.py

```diff
@@ -0,0 +1,246 @@
+import asyncio
+import threading
+from collections import defaultdict
+from functools import partial
+from itertools import groupby
+from typing import (
+    Any,
+    Awaitable,
+    Callable,
+    DefaultDict,
+    Dict,
+    List,
+    Mapping,
+    Optional,
+    Type,
+    TypeVar,
+    Union,
+)
+
+from langchain_core.runnables.base import (
+    Runnable,
+    RunnableSerializable,
+    coerce_to_runnable,
+)
+from langchain_core.runnables.config import RunnableConfig, patch_config
+from langchain_core.runnables.utils import ConfigurableFieldSpec, Input, Output
+
+T = TypeVar("T")
+Values = Dict[Union[asyncio.Event, threading.Event], Any]
+CONTEXT_CONFIG_PREFIX = "__context__/"
+CONTEXT_CONFIG_SUFFIX_GET = "/get"
+CONTEXT_CONFIG_SUFFIX_SET = "/set"
+
+
+async def _asetter(done: asyncio.Event, values: Values, value: T) -> T:
+    values[done] = value
+    done.set()
+    return value
+
+
+async def _agetter(done: asyncio.Event, values: Values) -> Any:
+    await done.wait()
+    return values[done]
+
+
+def _setter(done: threading.Event, values: Values, value: T) -> T:
+    values[done] = value
+    done.set()
```

Comment:
For these, it looks like this is used for the use case with parallel chains accessing the same context, is that a use case? In that case we could get into some sort of deadlock if it's the wrong order?

Reply:
YOLO